### PR TITLE
[Docs] Add decision record for core language strategy

### PR DESCRIPTION
**Pull request by foundrytom**
Part of #181, documents the high-level strategy for multi-language support. Specifics around the conventions we will use on the ground will come at a later date, once this choice is ratified.

Thanks to @feltech for the comprehensive code sketch.

## Notes for reviewers

The most digestible view of the document is [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/files?short_path=13a4af3#diff-13a4af328cd7c2e775857a05460352d18efd46f463288a24f7a158495bd2d980).

This PR serves to validate the decision itself, not just the correctness of its writeup (given it's me, I'm sure there will be some copy gremlins 🙃). Any parties with comment on the choice itself should feel free to use this to discuss the topic.

### CODE REVIEW

File: decisions/DR004-Core-language-support.md
Comment (feltech) [inline decisions/DR004-Core-language-support.md:90]: [consider] No C++ ABI isolation, so subject to compiler and library ABI breaks between the host, OpenAssetIO and manager plugins.

File: decisions/DR004-Core-language-support.md
Comment (feltech) [inline decisions/DR004-Core-language-support.md:103]: [consider] Presents a stable ABI isolation layer between hosts and plugins, so that the OpenAssetIO core, and manager plugins, are less likely to require recompilation when compilers and libraries are updated in host applications.

**Review by feltech — CHANGES_REQUESTED** [review]

File: decisions/DR004-Core-language-support.md
Comment (foundrytom) [reply decisions/DR004-Core-language-support.md:129]: Is that a con of the choice here, or just a con of that deployment scenario? It feels like more the former? I think we need to treat that scenario as a requirement, rather than an option, so it should probably show up as a con for options 2 instead - i.e. the addition of boilerplate to the happy path.

File: decisions/DR004-Core-language-support.md
Comment (foundrytom) [reply decisions/DR004-Core-language-support.md:112]: Good call, added [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/a3083492b088223f161091c09b156980224215bb).

File: decisions/DR004-Core-language-support.md
Comment (foundrytom) [reply decisions/DR004-Core-language-support.md:145]: Added a note that Option 3 natively supports the high-perf path [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/5408c60dc1ec565784c68318eff67bd22701d105), and added a note on the potential overhead in Opt 2 [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/5e678a5e6925b1a07500d21264baf2954518f70d).

File: decisions/DR004-Core-language-support.md
Comment (foundrytom) [reply decisions/DR004-Core-language-support.md:90]: Added [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/ff11aeee4d2e53abf1b3ba5a2e336b354952bbf6).

File: decisions/DR004-Core-language-support.md
Comment (foundrytom) [reply decisions/DR004-Core-language-support.md:106]: Added [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/73524625a3debd24b95c4827279b2b2c25395361).

**Comment by foundrytom** [comment]
@feltech thanks for extra points, I've added those, and rejigged a few of the existing ones to better reflect the impact of the other options on the two noted topics of ABI/perf. See what you think.

File: decisions/DR004-Core-language-support.md
Comment (meshula) [inline decisions/DR004-Core-language-support.md:118]: I disagree on the "significant coding" or "third-party depencies" point, and the subsequent risk. hash tables are well known solutions for maps, and it's not hard to design an `optional` surrogate. I'd word this point instead as 

The data model of OpenAssetIO _requires_ `map` and `optional` types. The lack of standardization in C around these concepts means a local implementation of these for C is required, and a mapping from this implementation to C++ and Python concepts introduces runtime and maintenance overhead.

File: decisions/DR004-Core-language-support.md
Comment (meshula) [inline decisions/DR004-Core-language-support.md:122]: C++ does not introduce safety to memory management. It only introduces some idioms, that can be used, to mitigate issues, but foot-guns abound.

File: decisions/DR004-Core-language-support.md
Comment (meshula) [inline decisions/DR004-Core-language-support.md:125]: An earlier point in this document is that Foundry has prior art in C function suites, this point seems contradictory. The advertised lack of generic programmings stems from a rejection by the C++ community of macros; perhaps a better wording would be:

C programming idioms around inheritance and generic programming use macros. The use of macros makes for awkward, or impossible interoperation with languages such as C++ where the use of such constructs interfere with clarity and best practices.

File: decisions/DR004-Core-language-support.md
Comment (meshula) [inline decisions/DR004-Core-language-support.md:172]: Some words about memory management are perhaps necessary. It's not sufficient to simply isolate the compiler. Across the C boundary, there should probably be some consideration that any objects obtained from the interface cannot be externally memory managed. In other words, the objects should be explicitly incompatible with malloc, free, realloc, and the string functions that reallocate.

File: decisions/DR004-Core-language-support.md
Comment (meshula) [inline decisions/DR004-Core-language-support.md:179]: my previous point being support of this point, of course, but trying to lay it out explicitly

File: decisions/DR004-Core-language-support.md
Comment (meshula) [inline decisions/DR004-Core-language-support.md:326]: years of fighting MSVC and string deduplication, and how to decorate extern statics for __declspec(dllexport) make me strongly prefer strings not being inlined like this.

I'd go so far as to say that both char* and std::string are more or less anti-patterns... Moving forward, I'm using the pattern of library specific string classes, with c_str and std::string adaptors.

the reason is that we generally want to impose constraints on allowable strings, such as notions of being well formed. all that goes out the window with the standard types, and all the many ways one can introduce a malformed string into the system become an ongoing maintenance concern.

A typical example is a string to represent a URL. If it's a string, it has to be validated and sanitized at every point of use. If it's a specialized value, it's sanitized/validated on construction, and can subsequently make a promise of being well formed.

File: decisions/DR004-Core-language-support.md
Comment (meshula) [inline decisions/DR004-Core-language-support.md:368]: so like here. I'm sure there's a notion of a well formed terminology string, and a malformed one.

File: decisions/DR004-Core-language-support.md
Comment (luke-titley) [reply decisions/DR004-Core-language-support.md:90]: A random thought... not sure if this is useful for you guys, but the rust wg wrote a tool (cppmm) which generates a c api for a c++ library. It's what we're using to make rust bindings.

File: decisions/DR004-Core-language-support.md
Comment (luke-titley) [reply decisions/DR004-Core-language-support.md:326]: +1 to your own string type. It doesn't actually have to support may operations if it can convert to/from std::string and char* and it gives you compiler independence for movability, alignment and size.

File: decisions/DR004-Core-language-support.md
Comment (luke-titley) [reply decisions/DR004-Core-language-support.md:90]: An example of usd stage in c. It's fairly easy to use and the whole project is in c++.

https://github.com/vfx-rs/usd-bind/blob/main/usd-c/include.in/pxr/usd/usd/usd_stage.h

File: decisions/DR004-Core-language-support.md
Comment (feltech) [reply decisions/DR004-Core-language-support.md:125]: > An earlier point in this document is that Foundry has prior art in C function suites, this point seems contradictory
Foundry's C function pointer suites (structs) invariably call through to C++ logic. We could clarify that point if its causing confusion.

File: decisions/DR004-Core-language-support.md
Comment (foundrytom) [reply decisions/DR004-Core-language-support.md:90]: Thanks @luke-titley looks really promising, we'll def factor this in when we start on the `C` exposed code for-real.

File: decisions/DR004-Core-language-support.md
Comment (foundrytom) [reply decisions/DR004-Core-language-support.md]: Reworded [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/8c053b05bd8dda23fb133640698a7900b2401514), also keeps the risk part, which is certainly a consideration from our p.o.v.

File: decisions/DR004-Core-language-support.md
Comment (foundrytom) [reply decisions/DR004-Core-language-support.md]: Re-worded this [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/153040cc29363227f3a6c9fcb81731854dec7f22), to indicate that it is not always infallible checks.

File: decisions/DR004-Core-language-support.md
Comment (foundrytom) [reply decisions/DR004-Core-language-support.md:126]: Added a little clarification to the section on prior-art [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/175d30a3194aad501b825bb0fd02750d02588de4), and broadened to include memory isolation [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/5c28736890dffce8644ef24127afcec60421130f).

File: decisions/DR004-Core-language-support.md
Comment (foundrytom) [reply decisions/DR004-Core-language-support.md]: Broadened the text to include memory [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/5c28736890dffce8644ef24127afcec60421130f). We will definitely factory in the `malloc` interaction patterns when we finalise the actual approach to strings/etc. in future work.

File: decisions/DR004-Core-language-support.md
Comment (foundrytom) [reply decisions/DR004-Core-language-support.md]: Updated the "cons" of a complex `C` implementation [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/6111631839f7432e8fade7434e28e778be8203ed).

**Comment by foundrytom** [comment]
Many thanks for all the input @meshula, @luke-titley. I think we've addressed relevant comments as noted. 

We have a general policy that a PR is blocked until review comments are resolved as satisfied by their author. I believe GitHub's permission mechanism means that you probably can't resolve your own conversation. If so, please add a 👍 as a reaction if you're happy to consider the thread resolved. We can close them for you and continue with the merge.

 @feltech, I've also conformed some formatting gremlins [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/8efd7e465f2b55f0cc30afed7e814fc7d0a8470b) and [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/4ffe4419845d1796fbf6fc365c9ea3df4d5e09a9).

File: decisions/DR004-Core-language-support.md
Comment (feltech) [inline decisions/DR004-Core-language-support.md:122]: [fix] Elsewhere programming languages are surrounded in backticks, but not "Rust" here.

**Review by feltech — APPROVED** [review]

File: decisions/DR004-Core-language-support.md
Comment (foundrytom) [reply decisions/DR004-Core-language-support.md]: Lol, no, good spot, thanks. Removed [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/d337559264006c1eb061176d56b82ee77a950617).

Comment (foundrytom) [reply decisions/DR004-Core-language-support.md]: Should be consistent [now](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/b829e705dbea283d8ffd98f06294a71ffc031bd5).

**Review by feltech — APPROVED** [review]
LGTM squash and merge when ready

**Pull request merged**
